syntax = "proto3";
package csce438;
import "google/protobuf/timestamp.proto";

// ------------------------------------------------------------
// The coordinator service definition
// These are the different functionalities exposed by service
// ------------------------------------------------------------

//Init and Heartbeat potentially redundant
service CoordService{
  rpc Heartbeat (ServerInfo) returns (Confirmation) {}
  rpc GetServer (ID) returns (ServerInfo) {}
  rpc GetSlaveInfo (ID) returns (ServerInfo) {}
  rpc GetUsersAddrs (UserList) returns (AddressList) {}
  // ZooKeeper API here
}

// message to ask coordinator about the synchronizer's address of other UserList
message UserList{
  repeated int32 users = 1;
}

message AddressInfo{
  string syncAddress = 1;
  int32 userId = 2;
}

message AddressList{
  repeated AddressInfo addressInfo = 1;
}

//server info message definition
message ServerInfo{
  int32 serverID = 1;
  string hostname = 2;
  string port = 3;
  string type = 4;
  int32 clusterID = 5;
}

//confirmation message definition
message Confirmation{
  bool status = 1;
  string type = 2;
}

//id message definition
message ID{
  int32 id = 1;
}


